packages/flutter_tools/lib/src/fuchsia/fuchsia_kernel_compiler.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../artifacts.dart';
+import '../base/common.dart';
+import '../base/context.dart';
+import '../base/file_system.dart';
+import '../base/io.dart';
+import '../base/logger.dart';
+import '../base/process_manager.dart';
+import '../build_info.dart';
+import '../convert.dart';
+import '../globals.dart';
+import '../project.dart';
+
+import 'fuchsia_sdk.dart';
+
+/// The [FuchsiaKernelCompiler] instance.
+FuchsiaKernelCompiler get fuchsiaKernelCompiler =>
+    context.get<FuchsiaKernelCompiler>();
+
+class FuchsiaKernelCompiler {
+  /// Compiles the [fuchsiaProject] with entrypoint [target] to a collection of
+  /// .dilp files, and a manifest that refers to them.
+  Future<void> build({
+    @required FuchsiaProject fuchsiaProject,
+    @required String target, // E.g., lib/main.dart
+    BuildInfo buildInfo = BuildInfo.debug,
+  }) async {
+    const String multiRootScheme = 'main-root';
+    final String packagesFile = fuchsiaProject.project.packagesFile.path;
+    final String outDir = getFuchsiaBuildDirectory();
+    final String appName = fuchsiaProject.project.manifest.appName;
+    final String fsRoot = fuchsiaProject.project.directory.path;
+    final String relativePackagesFile =
+        fs.path.relative(packagesFile, from: fsRoot);
+    final String manifestPath = fs.path.join(outDir, '$appName.dilpmanifest');
+    List<String> flags = <String>[
+      // https://github.com/dart-lang/sdk/issues/36639:
+      // Remove when new constant eval supports dilp files.
+      '--enable-experiment=no-constant-update-2018',
+      '--target', 'flutter_runner',
+      '--platform', fuchsiaArtifacts.platformKernelDill.path,
+      '--filesystem-scheme', 'main-root',
+      '--filesystem-root', fsRoot,
+      '--packages', '$multiRootScheme:///$relativePackagesFile',
+      '--output', fs.path.join(outDir, '$appName.dil'),
+      '--no-link-platform',
+      '--split-output-by-packages',
+      '--manifest', manifestPath,
+      '--component-name', appName,
+    ];
+
+    if (buildInfo.isDebug) {
+      flags += <String>[
+        '--embed-sources',
+      ];
+    } else if (buildInfo.isProfile) {
+      flags += <String>[
+        '--no-embed-sources',
+        '-Ddart.vm.profile=true',
+        '--gen-bytecode',
+        '--drop-ast',
+      ];
+    } else if (buildInfo.isRelease) {
+      flags += <String>[
+        '--no-embed-sources',
+        '-Ddart.vm.release=true',
+        '--gen-bytecode',
+        '--drop-ast',
+      ];
+    } else {
+      throwToolExit('Expected build type to be debug, profile, or release');
+    }
+
+    flags += <String>[
+      '$multiRootScheme:///$target',
+    ];
+
+    final List<String> command = <String>[
```

Comment:
The formatting here is a bit awkward, maybe:

```
final List<String> command = <String>[
  artifacts.getArtifactPath(Artifact.engineDartBinary),
  fuchsiaArtifacts.kernelCompiler.path,
  ..flags,
]
```

Or

```
final List<String> command = <String>[
  artifacts.getArtifactPath(Artifact.engineDartBinary),
  fuchsiaArtifacts.kernelCompiler.path,
]..addAll(flags);
```

Reply:
Done